Update the game logic to handle the game board and associated pieces and squares

ALHennings opened the pull request:
This is a fairly large update, which should provide the backbone for the game mechanics. It is under development, but functional. 

Closes issues #104, #105, #106

ALHennings committed:
Testing Implementation of Junit

Aded JungleGame and TestJungleGame working on Junit

ALHennings committed:
Put current files in a sub folder

While I figure out the system and the correct balance, I am placing the game files in a sub folder. This can be reversed without major complication at any point.

ALHennings committed:
Adding GamePiece class and enum

ALHennings committed:
Updates for initial system.

Current Updats include a player class, a boardsquare class, and a gamepiece class (with associated enumeration).

ALHennings committed:
Merge branch 'development' into p3/Alexander/initialDevelopment

ALHennings committed:
Update junglegame

ALHennings committed:
Update class path

ALHennings committed:
Merge branch 'development' into p3/Alexander/initialDevelopment

ALHennings committed:
Updating the system to have a gameboard and associated pieces

ALHennings committed:
Updated Gameboard to handle player pieces

Also setting up a new array feature for the board.

ALHennings committed:
Using playerColor for Board Square

Added another type to the enumeration fo PlayerColor

ALHennings committed:
Update Gameboard with full set of Squares

ALHennings committed:
Update Pieces to include color

ALHennings committed:
Updates for Gameboard, further disccusion needed before moving forward.

ALHennings committed:
Updating the JungleGame

Otherwise this will be left alone until I have a chance to figure it out amongst the team. Currently this is the interaction point between the "game" and both the ui and server.

ALHennings committed:
Merge branch 'development' into p3/Alexander/initialDevelopment

ALHennings committed:
Set up a beginning check move

This is a complicate method it will take some time.

ALHennings committed:
Added to check move

Added the check for type of piece

ALHennings committed:
Finish checkMove

Check move should now be functional, with this most of the movement should fall into place. We need to check interfaces, and checkMove NEEDS TESTED.

ALHennings committed:
Added trap control to move check

ALHennings committed:
update comment

ALHennings committed:
Edit checkMove

Edit checkMove to handle a rat in the water when jumping the river

ALHennings committed:
Update Comments and Check for dead piece

ALHennings committed:
Merge branch 'development' into p3/Alexander/initialDevelopment

ALHennings committed:
Update irregular arguments

tking2096 commented on client/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleClient/app/Game/BoardSquare.java:
```diff
@@ -0,0 +1,24 @@
+package edu.colostate.cs.cs414.chesshireCoders.jungleClient.app.Game;
+
+public class BoardSquare {
+	SquareType typeOfSquare = SquareType.Normal;
```
Shouldn't this be taken care of with inheritance / polymorphism?

ALHennings committed:
Update Check move with better arguments.

Changed the piece int in checkMove from the array value (aka Rat = 0) to the power level of the piece (aka Rat = 1)

tking2096 commented on client/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleClient/app/Game/GamePiece.java:
```diff
@@ -0,0 +1,37 @@
+package edu.colostate.cs.cs414.chesshireCoders.jungleClient.app.Game;
+
+public class GamePiece {
+	PieceType typeOfPiece = PieceType.Rat;
```
Shouldn't this also be taken care of with inheritance / polymorphism?

tking2096 commented on client/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleClient/app/Game/BoardSquare.java:
```diff
@@ -0,0 +1,24 @@
+package edu.colostate.cs.cs414.chesshireCoders.jungleClient.app.Game;
```
I think we should keep our package names lowercase. This looks like a class `Game` that is in the `app` package.

ALHennings replied:
Fixed.

tking2096 commented on client/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleClient/app/Game/GameBoard.java:
```diff
@@ -0,0 +1,275 @@
+	//piece must be the powerlevel of the piece (aka Rat = 1 NOT 0)
+	//column = X axis location
+	//row = Y Axis Location
+	public boolean checkMove(int color, int piece, int column, int row) {		
```
I think `isValidMove()` would be a better name for readability. Also this method is very long. You can have a method for each of these checks.

ALHennings replied:
Fixed.

ALHennings committed:
Preventing future error

tking2096 commented on client/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleClient/app/Game/GameBoard.java:
```diff
@@ -0,0 +1,275 @@
+		
+		//The only time the row or column will increase by more than one is if the tiger or leopard is jumping the river
+		//Check if it increases by more than one
+		if(!(((pieceColumn+1 == column)||(pieceColumn-1 == column))||((pieceRow+1 == row)||(pieceRow-1 == row)))){
```
This conditional is hard to read. Consider splitting it up or replacing clauses with method calls.

ALHennings replied:
All it does is check if it's 1 higher or lower. I'm not sure how to make that any easier. I've made each separate logic encapsulated in the brackets.

tking2096 commented on client/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleClient/app/Game/GameBoard.java:
```diff
@@ -0,0 +1,275 @@
+		//Check if it increases by more than one
+		if(!(((pieceColumn+1 == column)||(pieceColumn-1 == column))||((pieceRow+1 == row)||(pieceRow-1 == row)))){
+			//if it is not a tiger or a leopard return false
+			if(!((piece == 4)||(piece == 5))) {
```
This method depends on the type of the piece. Maybe it would be better to put it within the `GamePiece` classes and have polymorphism handle it. That way it would just be a call to `piece.isValidMove(row, column)` or `piece.canMoveTo(row, column)`.

ALHennings replied:
I can see the argument here, however as I can see the code will remain the same, it can be moved if need be, but it's the same either way, and this way reduces coupling while keeping the board design in once place.

tking2096 commented on client/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleClient/app/Game/GameBoard.java:
```diff
@@ -0,0 +1,275 @@
+				return false;
+			}
+			//if it is not jumping the left river it must jump right.
+			if(!(((pieceColumn == 0) && (column == 3)) || ((pieceColumn == 3) && (column == 0)) || ((pieceRow == 2) && (row == 6)) || ((pieceRow == 6) && (row == 2)))) {
```
Again, hard to read.

tking2096 commented on client/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleClient/app/Game/GameBoard.java:
```diff
@@ -0,0 +1,275 @@
+		BoardSquare boardSquare = boardSquares[column][row];
+		BoardSquare pieceSquare = boardSquares[gamePiece.getColumn()][gamePiece.getRow()];
+		boolean enemyBanned = false;
+		if(boardSquare.getSquareType().equals(SquareType.River)) {
```
You could handle this with polymorphism with something like `piece.canOccupy(BoardSquare)`

ALHennings replied:
This will again increase coupling while using the same code in a different spot.

tking2096 commented on client/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleClient/app/Game/GameBoard.java:
```diff
@@ -0,0 +1,275 @@
+		
+		//Check if their is a creature on it with a higher power
+		//if any of the player's pieces are on the space return false;
+		for(int i = 0; i < 8; i++) {
```
Can't you just get the `GamePiece` at `(row, column)` and compare it's power with `piece`?

ALHennings replied:
That's what that does? the comments have been fixed to clarify.

tking2096 commented on client/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleClient/app/Game/GameBoard.java:
```diff
@@ -0,0 +1,275 @@
+		
+		//if the space the piece is moved to is a FRIENDLY trap, any enemy will be consumed so just return true;
+		//is it a trap?
+		if (boardSquare.getSquareType().equals(SquareType.Trap)){
```
This could be handled by the previous check. We just need to set the power of a `GamePiece` to 1 when it enters a trap and `piece.setPowerDefault()` when it exits.

ALHennings replied:
I can either make a method to control the current power of a piece  requiring an update each move, or I can have one check in this one method. It seemed simpler.

tking2096 commented on client/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleClient/app/Game/PieceType.java:
```diff
@@ -0,0 +1,5 @@
+package edu.colostate.cs.cs414.chesshireCoders.jungleClient.app.Game;
+
+public enum PieceType {
```
Not necessary with polymorphism?

tking2096 commented on client/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleClient/app/Game/SquareType.java:
```diff
@@ -0,0 +1,5 @@
+package edu.colostate.cs.cs414.chesshireCoders.jungleClient.app.Game;
+
+public enum SquareType {
```
Not necessary with polymorphism?

tking2096 reviewed (changes requested):
Please excuse the short comments, I don't mean them to sound rude.

amgunden commented on client/src/main/java/edu/colostate/cs/cs414/chesshireCoders/jungleClient/app/Game/Player.java:
```diff
@@ -0,0 +1,11 @@
+	//I'm not sure what this value is supposed to be so I'm changing user to userID
+	int userID = 0;
+
+	public Player() {
```
For the Player class, could you add the color and userID as parameters for the constructor and have it set them

ALHennings replied:
I have not done player yet, as noted this is only for the board game, the squares, and the pieces. Both the game controller and player are not finished/used.

ALHennings committed:
Fixing Game = game. Fixing long method.

Changed package name.
Broke the isValidMove method into smaller portions.

ALHennings committed:
Fix error

ALHennings committed:
Update Player constructor

Added on Adam's request. Player Unfinished.

ALHennings committed:
Update Missing case
